File: packages/ts/src/env.test.ts
import { sep } from 'node:path'
import { describe, expect, it } from 'vitest'

import { coerceValue, envKeyPath, envLayer } from './env.js'
import type { Layer } from './merge.js'
import { buildSources } from './sources.js'
import { warningCollector } from './test-utils.js'

describe('coercion — SPEC §4.5 step 5', () => {
  it.each([
    { raw: '5432', expected: 5432 as unknown, type: 'number' },
    { raw: 'true', expected: true as unknown, type: 'boolean' },
    { raw: '[1,2]', expected: [1, 2] as unknown, type: 'array' },
    { raw: '5432abc', expected: '5432abc' as unknown, type: 'string' },
  ])('$raw becomes the $type', ({ raw, expected }) => {
    expect(coerceValue(raw)).toEqual(expected)
  })

  it('keeps a value the caller says was quoted as a string', () => {
    expect(coerceValue('5432', true)).toBe('5432')
  })
})

describe('envKeyPath', () => {
  it('splits on __ and leaves a single underscore literal', () => {
    expect(envKeyPath('LOG__LEVEL')).toEqual(['log', 'level'])
    expect(envKeyPath('SOME_KEY')).toEqual(['some_key'])
    expect(envKeyPath('A__B__C')).toEqual(['a', 'b', 'c'])
  })
})

describe('envLayer', () => {
  it('maps MYTOOL_LOG__LEVEL to a nested key', () => {
    expect(envLayer({ MYTOOL_LOG__LEVEL: 'trace' }, 'MYTOOL')).toEqual({ log: { level: 'trace' } })
  })

  it('keeps a single underscore inside the key name', () => {
    expect(envLayer({ MYTOOL_SOME_KEY: '1' }, 'MYTOOL')).toEqual({ some_key: 1 })
  })

  it('ignores variables without the prefix', () => {
    expect(envLayer({ OTHER_LOG__LEVEL: 'trace', PATH: '/usr/bin' }, 'MYTOOL')).toEqual({})
  })

  it('warns about a variable that maps to an empty key path', () => {
    const { warnings, onWarning } = warningCollector()
    expect(envLayer({ MYTOOL_: 'x' }, 'MYTOOL', onWarning)).toEqual({})
    expect(warnings[0]).toMatch(/empty key path/)
  })
})

describe('buildSources', () => {
  const layerAt = (path: string, value: unknown, precedence: number): Layer => ({
    value,
    source: { path, format: 'toml', precedence, keys: [] },
  })

  it('reports the top-level keys each source contributed, sorted', () => {
    const sources = buildSources([layerAt('/abs/a/config.toml', { b: 1, a: 2 }, 2)])
    expect(sources[0]?.keys).toEqual(['a', 'b'])
  })

  it('includes a source whose file parsed empty, with no keys', () => {
    expect(buildSources([layerAt('/abs/a/config.toml', {}, 2)])[0]?.keys).toEqual([])
  })

  it('rewrites paths relative to relativeTo, forward-slashed', () => {
    const root = ['', 'abs', 'fixture'].join(sep)
    const file = ['', 'abs', 'fixture', 'project', '.config', 'mytool', 'config.toml'].join(sep)
    expect(buildSources([layerAt(file, {}, 2)], { relativeTo: root })[0]?.path).toBe(
      'project/.config/mytool/config.toml',
    )
  })

  it('leaves the non-file layer labels alone', () => {
    const sources = buildSources([{ value: { a: 1 }, source: { path: '<env>', format: 'env', precedence: 4, keys: [] } }], {
      relativeTo: '/abs/fixture',
    })
    expect(sources[0]?.path).toBe('<env>')
  })

  it('preserves application order rather than sorting by precedence', () => {
    // A user-level .env is precedence 3 but belongs inside the user root's block, so it must
    // still be emitted before the project files that outrank it (SPEC §3.1).
    const order = buildSources([
      layerAt('/u/config.toml', {}, 1),
      layerAt('/u/.env', {}, 3),
      layerAt('/p/config.toml', {}, 2),
    ]).map((source) => source.precedence)
    expect(order).toEqual([1, 3, 2])
  })
})

File: packages/golang/env_test.go
package configdiscovery

import (
	"encoding/json"
	"path/filepath"
	"reflect"
	"testing"
)

func TestEnvCoercion(t *testing.T) {
	cases := []struct {
		raw  string
		want any
	}{
		{"5432", int64(5432)},
		{"true", true},
		{"[1,2]", []any{int64(1), int64(2)}},
		{"5432abc", "5432abc"},
		{"1.5", 1.5},
	}
	for _, testCase := range cases {
		if got := coerceValue(testCase.raw, false); !reflect.DeepEqual(got, testCase.want) {
			t.Fatalf("%q: got %#v, want %#v", testCase.raw, got, testCase.want)
		}
	}
	if got := coerceValue("5432", true); got != "5432" {
		t.Fatalf("a quoted value must stay a string, got %#v", got)
	}
}

func TestEnvLayerMapping(t *testing.T) {
	cases := []struct {
		name string
		env  map[string]string
		want map[string]any
	}{
		{
			name: "__ splits nesting",
			env:  map[string]string{"MYTOOL_LOG__LEVEL": "trace"},
			want: map[string]any{"log": map[string]any{"level": "trace"}},
		},
		{
			name: "a single underscore stays literal",
			env:  map[string]string{"MYTOOL_SOME_KEY": "1"},
			want: map[string]any{"some_key": int64(1)},
		},
		{
			name: "another prefix is ignored",
			env:  map[string]string{"OTHER_LOG__LEVEL": "trace", "PATH": "/usr/bin"},
			want: map[string]any{},
		},
	}
	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			if got := envLayer(testCase.env, "MYTOOL", nil); !reflect.DeepEqual(got, testCase.want) {
				t.Fatalf("got %#v, want %#v", got, testCase.want)
			}
		})
	}
}

func TestEnvLayerWarnsOnAnEmptyKeyPath(t *testing.T) {
	var warnings []string
	envLayer(map[string]string{"MYTOOL_": "x"}, "MYTOOL", func(m string) { warnings = append(warnings, m) })
	if len(warnings) != 1 {
		t.Fatalf("expected a warning, got %v", warnings)
	}
}

func TestSourcesKeysAndOrder(t *testing.T) {
	sources := buildSources([]Layer{
		{Value: map[string]any{"b": 1, "a": 2}, Source: Source{Path: "/u/config.toml", Format: "toml", Precedence: 1}},
		{Value: map[string]any{}, Source: Source{Path: "/u/.env", Format: "dotenv", Precedence: 3}},
		{Value: map[string]any{"c": 3}, Source: Source{Path: "/p/config.toml", Format: "toml", Precedence: 2}},
	}, "")

	if !reflect.DeepEqual(sources[0].Keys, []string{"a", "b"}) {
		t.Fatalf("keys: got %v", sources[0].Keys)
	}
	// A file that parsed empty is still reported, with no keys.
	if len(sources[1].Keys) != 0 {
		t.Fatalf("expected an empty key list, got %v", sources[1].Keys)
	}
	// Application order, not sorted by precedence: a user-level .env belongs inside the user
	// root's block even though it outranks that root's structured files (SPEC section 3.1).
	got := []int{sources[0].Precedence, sources[1].Precedence, sources[2].Precedence}
	if !reflect.DeepEqual(got, []int{1, 3, 2}) {
		t.Fatalf("order: got %v", got)
	}
}

func TestSourcesEmptyKeysMarshalAsArrayNotNull(t *testing.T) {
	sources := buildSources([]Layer{
		{Value: map[string]any{}, Source: Source{Path: "/u/config.toml", Format: "toml", Precedence: 1}},
	}, "")
	encoded, err := json.Marshal(sources)
	if err != nil {
		t.Fatal(err)
	}
	want := `[{"path":"/u/config.toml","format":"toml","precedence":1,"keys":[]}]`
	if string(encoded) != want {
		t.Fatalf("got %s\nwant %s", encoded, want)
	}
}

func TestSourcesRelativeTo(t *testing.T) {
	root := filepath.Join(string(filepath.Separator), "abs", "fixture")
	file := filepath.Join(root, "project", ".config", "mytool", "config.toml")
	sources := buildSources([]Layer{
		{Value: map[string]any{}, Source: Source{Path: file, Format: "toml", Precedence: 2}},
		{Value: map[string]any{"a": 1}, Source: Source{Path: "<env>", Format: "env", Precedence: 4}},
	}, root)

	if sources[0].Path != "project/.config/mytool/config.toml" {
		t.Fatalf("got %q", sources[0].Path)
	}
	// The labels are not paths and are passed through untouched.
	if sources[1].Path != "<env>" {
		t.Fatalf("got %q", sources[1].Path)
	}
}
